### DISCUSSION TITLE
Can`t force DependencyConvergence rule to work

### DISCUSSION THREAD

**Discussion by pshakhov** [kordamp/enforcer-gradle-plugin#31] [Q&A] [answered]
Hello! I have a problem with usage of DependencyConvergence rule.

settings.gradle.kts:
```
buildscript {
    repositories {
        gradlePluginPortal()
    }
    dependencies {
        classpath("org.kordamp.gradle:enforcer-gradle-plugin:0.13.0")
    }
}
apply(plugin = "org.kordamp.gradle.enforcer")

rootProject.name = "enforcer-gragle-plugin-example"
```

build.gradle.kts:
```
import enforcer.rules.BannedDependencies
import enforcer.rules.DependencyConvergence
import org.kordamp.gradle.plugin.enforcer.api.EnforcerLevel

plugins {
    `java-library`

    id("org.kordamp.gradle.project-enforcer")
}

group = "org.example"
version = "1.0-SNAPSHOT"

repositories {
    mavenCentral()
}

apply(plugin = "org.kordamp.gradle.project-enforcer")

enforce {

    // can`t force to work
  rule(DependencyConvergence::class.java) {
      failOnDynamicVersions.set(true)
      failOnChangingVersions.set(true)
      failOnNonReproducibleResolution.set(true)
  }

  // works as expected
  rule(BannedDependencies::class.java) {
      enforcerLevel.set(EnforcerLevel.WARN)
      exclude("org.slf4j:slf4j-api")
  }
}

dependencies {
    testImplementation(platform("org.junit:junit-bom:5.9.1"))
    testImplementation("org.junit.jupiter:junit-jupiter")

    implementation("org.slf4j:slf4j-jdk14:1.6.1")
    implementation("org.slf4j:slf4j-nop:1.6.0")
}

tasks {
    test {
        useJUnitPlatform()
    }

    wrapper {
        gradleVersion = "6.8.2"
    }
}
```

Getting log for BannedDependencies rule:
`
> Configure project :
[project-enforcer :] [AFTER_PROJECT :] An Enforcer rule has failed
org.kordamp.gradle.plugin.enforcer.api.EnforcerRuleException: 
Enforcer rule 'enforcer.rules.BannedDependencies' was triggered.
Found Banned Dependency: org.slf4j:slf4j-api:1.6.1 in project :
Disable this rule temporarily with -Denforcer.rules.BannedDependencies.enabled=false and
invoke 'dependencies' to locate the source of the banned dependencies.
`

but if disable it and use only DependencyConvergence rule or both configuration not fails.

What Am I Doing Wrong? Can push minimal, reproducible Example if needed.

**Comment by aalmiray** [comment]
FWIW you don't have to configure both enforcer plugins. For what I can see use only the `project-enforcer` for the time being.

**Comment by pshakhov** [comment]
`enforce` extension used like in doc [example](https://kordamp.org/enforcer-gradle-plugin/#_dsl_configure_build) can`t resolve in settings file:
<img width="689" alt="image" src="https://github.com/kordamp/enforcer-gradle-plugin/assets/32340504/ad7ffa88-1229-45e5-9874-28e158197cf8">

**Reply by aalmiray** [reply]
This plugin configures itself, showing that then `enforce` config section is indeed available in `settings.gradle` https://github.com/kordamp/kordamp-parent/blob/master/settings.gradle

**Reply by pshakhov** [reply]
You right, after then i used `settings.gradle` in Groovy instead kotlin script i`ve got:
```
Caused by: org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.conflicts.VersionConflictException: Conflict(s) found for the following module(s):
  - org.slf4j:slf4j-api between versions 1.6.1 and 1.6.0
```

But our projects used kts not Groovy, and i don`t understand how to configure it properly to working state.

**Reply by aalmiray** [reply]
If it works with Groovy and it doesn't with Kotlin then it could be:
- a limitation of Kotlin DSL in `settings.gradle.kts`
- missing metadata exposed by the plugin so that Kotlin DSL may work as expected

I don't use Kotlin at all thus I'm unable to tell which may be the actual cause of the problem. Groovy works fine and thus that's why I use it.

**Reply by aalmiray** [reply]
FWIW `enforce` is just like any other extension point exposed by a plugin. If additional syntax were to be required by the Kotlin DSL to use such extension points then that would be my first guess.

**Comment by pshakhov** [comment]
Can i add to readme way to configure enforcer using kotlin script which i realized?

**Reply by aalmiray** [reply]
You can write it down in this discussion topic.

**Reply by pshakhov** [reply, marked as answer]
``` 
pluginManagement {
    plugins {
        id("org.kordamp.gradle.project-enforcer")
    }
 
    repositories {
        mavenCentral()
        gradlePluginPortal()
    }
}
 
buildscript {
    repositories {
        mavenCentral()
        gradlePluginPortal()
    }
    dependencies {
        classpath("org.kordamp.gradle:enforcer-gradle-plugin:0.13.0")
    }
}
 
apply(plugin = "org.kordamp.gradle.enforcer")
 
configure<org.kordamp.gradle.plugin.enforcer.api.BuildEnforcerExtension> {}
```
